### PR TITLE
Fix fuzz tests with property-based testing and add CI

### LINKED ISSUE #3: Fix fuzz tests of mem
Use rust library for fuzzy testing. Fix broken tests, use more general and randomly testing (with indirect order of memory operations shrink/grow) with random sizes and relative results. Rewrite this file. Also remove all useless documentation from tests 

```
running 1 test
error: test failed, to rerun pass `-p mem --test fuzz`

Caused by:
  process didn't exit successfully: target\debug\deps\fuzz-73173ab4409e51f3 interleaved_ops --exact --nocapture` (exit code: 0xc0000374, STATUS_HEAP_CORRUPTION)

```

Write CI tests that runs tests for each crate of this package (from crates), or run the all tests btw

### PR DESCRIPTION
**Pull request by konard**
## Summary

This PR addresses issue #3 by rewriting the fuzz tests to use proper property-based testing with the `proptest` library and adding a comprehensive CI workflow.

### Fuzz Tests Rewrite
- ✅ Replaced manual test cases with property-based testing using `proptest`
- ✅ Added `random_alloc_ops` and `random_prealloc_ops` tests that generate random sequences of grow/shrink operations
- ✅ Added `grow_maintains_data` test to verify data integrity across multiple grow operations
- ✅ Added `shrink_preserves_remaining` test to verify data preservation after shrinking
- ✅ Tests now use random sizes and indirect ordering of memory operations as requested
- ✅ Kept essential tests for invariants, edge cases, and mutability
- ✅ Removed verbose documentation from test functions per issue requirements

### Test Organization
- ✅ Extracted general memory tests (`basic_invariants`, `edge_cases`, `mutability`) into `general.rs` module
- ✅ These tests are now applicable to all `RawMem` implementations via the `define_impls!` macro
- ✅ Moved `prealloc_overgrow` test to `pre.rs` module where it belongs

### CI Workflow
- ✅ Added `.github/workflows/ci.yml` that runs tests for all workspace crates
- ✅ Tests run on both stable and nightly Rust
- ✅ Includes format checking with `rustfmt` using **nightly toolchain** (for edition 2024 support)
- ✅ Includes linting with `clippy`

### Code Quality
- ✅ Updated `rustfmt.toml` to use **edition 2024** as requested
- ✅ Fixed all line length issues to comply with `max_width = 80`
- ✅ Fixed clippy warnings (unused imports, collapsible if statements)
- ✅ All tests passing locally
- ✅ No regressions in existing functionality

## Test Results

```
running 5 tests
test capacity_overflow ... ok
test grow_maintains_data ... ok
test random_prealloc_ops ... ok
test random_alloc_ops ... ok
test shrink_preserves_remaining ... ok

test result: ok. 5 passed; 0 failed

running 6 tests
test basic_invariants ... ok
test edge_cases ... ok
test mutability ... ok
test pre::ignore_grow_fillup ... ok
test pre::prealloc_overgrow ... ok
test miri ... ok

test result: ok. 6 passed; 0 failed
```

## Review Comments Addressed

1. ✅ Updated `rustfmt.toml` to use `edition = "2024"` instead of `version = "Two"`
2. ✅ Changed CI to use nightly toolchain for rustfmt (to support edition 2024 features)
3. ✅ Moved `prealloc_overgrow` test from `fuzz.rs` to `pre.rs` module
4. ✅ Extracted general tests to `general.rs` module, making them applicable to all `RawMem` implementations

Fixes #3

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    strategy:
+      matrix:
+        rust: [stable, nightly]
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{ matrix.rust }}
+      - name: Run tests for all workspace crates
+        run: cargo test --workspace --all-features
```
Comment (uselessgoddess) [inline .github/workflows/ci.yml]: i need to use this rustfmt.toml:
```toml
edition = "2024"
error_on_line_overflow = true
error_on_unformatted = true

imports_granularity = "One"
use_small_heuristics = "Max"
use_field_init_shorthand = true

max_width = 80
tab_spaces = 2
```
but there are some nightly features (but it is just formatter). We need to check fmt in nightly channel probably

**Comment by konard** [comment]
🤖 **AI Work Session Started**

Starting automated work session at 2025-12-02T08:34:54.788Z

The PR has been converted to draft mode while work is in progress.

_This comment marks the beginning of an AI work session. Please wait working session to finish, and provide your feedback._

**Comment by konard** [comment]
## ✅ Review Comments Addressed

All requested changes have been implemented:

### 1. Updated rustfmt configuration
- Changed `rustfmt.toml` from `version = "Two"` to `edition = "2024"`
- Updated CI workflow to use **nightly toolchain** for rustfmt checks (required for edition 2024 features)

### 2. Test reorganization
- ✅ Moved `prealloc_overgrow` test from `fuzz.rs` to `pre.rs` module (crates/mem/tests/fuzz.rs:163)
- ✅ Extracted general memory tests (`basic_invariants`, `edge_cases`, `mutability`) to new `general.rs` module (crates/mem/tests/fuzz.rs:146)
- These general tests are now applicable to all `RawMem` implementations through the `define_impls!` macro

### 3. Code quality improvements
- Fixed all line length violations (max_width = 80)
- All comments now comply with rustfmt requirements

### Local test results
All checks pass locally:
- ✅ `cargo test --workspace --all-features` - **14 tests passed**
- ✅ `cargo clippy --workspace --all-features -- -D warnings` - **no warnings**
- ✅ `cargo +nightly fmt --all -- --check` - **formatting correct**

The PR is now ready for re-review!

File: crates/mem/src/alloc.rs [hunk 2]
```diff
@@ -70,7 +70,8 @@ impl<T: Pod> RawMem for Alloc<T> {
       let old_layout =
         Layout::array::<T>(old_cap).map_err(|_| Error::CapacityOverflow)?;
 
-      // SAFETY: reallocating with matching old_layout and growing to larger size
+      // SAFETY: reallocating with matching old_layout and growing to
+      // larger size
```
Comment (uselessgoddess) [inline crates/mem/src/alloc.rs]: avoid this long words, prefer to use realloc, grow/shrink
SAFETY comments must me oneline if it is short

**Comment by konard** [comment]
## 🤖 Solution Draft Log
This log file contains the complete execution trace of the AI solution draft process.

💰 **Cost estimation:**
- Public pricing estimate: $0.123810 USD
- Calculated by Anthropic: $0.680326 USD
- Difference: $0.556516 (+449.49%)
📎 **Log file uploaded as GitHub Gist** (210KB)
🔗 [View complete solution draft log](https://gist.githubusercontent.com/konard/a8f1a661da42e6e2b1cda3af94bcee17/raw/a65012e5619f58e094eeed77792ea56667802e25/solution-draft-log-pr-1764666388309.txt)
---
*Now working session is ended, feel free to review and add any feedback on the solution draft.*
